api/lib/opentelemetry/resources/resource.rb

```diff
@@ -0,0 +1,24 @@
+# frozen_string_literal: true
+
+# Copyright 2019 OpenTelemetry Authors
+#
+# SPDX-License-Identifier: Apache-2.0
+
+module OpenTelemetry
+  module Resources
+    # Resource represents a resource, which captures identifying information about the entities
+    # for which telemetry (metrics or traces) is reported.
+    class Resource
+      attr_reader :labels
+
+      def initialize(**kvs)
+        # TODO: how defensive should we be here?
+        @labels = kvs.to_h { |k, v| [k.to_s.clone.freeze, v.to_s.clone.freeze] }.freeze
+      end
+
+      def merge(other)
+        new(labels.merge(other.labels))
```

Comment:
`self.class.new` or `Resource.new`. And we should probably check that the types are the same (i.e. this is a resource from the same implementation).